### PR TITLE
Frontend/initial components

### LINKED ISSUE #7: frontend: initial components
Add a first draft for a set of components required to render the data defined in #2 and #5.

### PR DESCRIPTION
**Pull request by ls505-hro**
Prototyp für das Design der Website hinzugefügt (zur Orientierung beim Entwickeln)
Componenten Post und Comment erstellt

### CODE REVIEW

File: frontend/.vscode/settings.json [hunk 1]
```diff
@@ -15,7 +15,7 @@
     "editor.formatOnSave": true
   },
   "[typescriptreact]": {
-    "editor.defaultFormatter": "esbenp.prettier-vscode",
+    "editor.defaultFormatter": "vscode.typescript-language-features",
```
Comment (marvindv) [inline frontend/.vscode/settings.json]: Did you have had problems with prettier? Or did VSCode ask which formatter to use?

File: frontend/src/App.tsx [hunk 2]
```diff
@@ -1,20 +1,20 @@
 import React from 'react';
-import logo from './logo.svg';
 import './App.css';
+import logo from './logo.svg';
```
Comment (marvindv) [inline frontend/src/App.tsx]: This change of import order probably happened because of the change in formatter. If there where no problems please change back to prettier and revert the changes on this line.

File: frontend/src/App.tsx [hunk 3]
```diff
@@ -1,20 +1,20 @@
...
 
 function App() {
   return (
-    <div className="App">
-      <header className="App-header">
-        <img src={logo} className="App-logo" alt="logo" />
+    <div className='App'>
```
Comment (marvindv) [inline frontend/src/App.tsx]: The changes here happened because I goofed up my previous changes, they where not formatted according to prettier rules. So we should just keep the changes in this and the following lines, even tough it makes the commit a bit more "noisy".

File: frontend/src/component/Post.tsx [hunk 4]
```diff
@@ -0,0 +1,34 @@
+import React, { useState } from 'react';
+import './Components.css';
```
Comment (marvindv) [inline frontend/src/component/Post.tsx]: I would suggest to put the styles for each individual component in different css files and import them in the corresponding source file of that component.

File: frontend/src/component/Post.tsx [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  name: string;
+  time: string;
+  content: string;
+  likes: number;
+  dislikes: number;
+  isComment: boolean;
+}
+
+function Post(props: PostProps) {
+  const [name] = useState<string>(props.name);
```
Comment (marvindv) [inline frontend/src/component/Post.tsx]: There is no need to put all those props into a state, especially since for none of them a setter is used. So either just reference the individual props with `props.time` and so on or maybe use destructuring like so:

```javascript
const { name, time, content, likes, dislikes } = props;
```

File: frontend/src/component/Post.tsx [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+  return (
+    <div className={className}>
+      <PostHeader name={name} time={time} isComment={props.isComment} />
+      <PostContent content={content} />
+      <PostFooter likes={likes} dislikes={dislikes} isComment={props.isComment} />
+    </div>
+  );
+}
+
+export default Post;
```
Comment (marvindv) [inline frontend/src/component/Post.tsx]: I would suggest reconsidering the decision to use default exports, as they in general don't really improve anything significantly. For example have a look at https://basarat.gitbook.io/typescript/main-1/defaultisbad

File: frontend/src/component/Post.tsx [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+function Post(props: PostProps) {
+  const [name] = useState<string>(props.name);
+  const [time] = useState<string>(props.time);
+  const [content] = useState<string>(props.content);
+  const [likes] = useState<number>(props.likes);
+  const [dislikes] = useState<number>(props.dislikes);
+
+  var className = props.isComment ? 'Comment' : '';
```
Comment (marvindv) [inline frontend/src/component/Post.tsx]: Use `const `. There is not really a reason to use `var` anymore.

File: frontend/src/component/PostContent.tsx [hunk 8]
```diff
@@ -0,0 +1,14 @@
+import React, { useState } from 'react';
+import './Components.css';
+
+interface PostContentProps {
+  content: string;
+}
+
+function PostContent(props: PostContentProps) {
+  const [content] = useState<string>(props.content);
```
Comment (marvindv) [inline frontend/src/component/PostContent.tsx]: Use props directly.

File: frontend/src/component/PostFooter.tsx [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+import './Components.css';
+
+interface PostFooterProps {
+  likes: number;
+  dislikes: number;
+  isComment: boolean;
+}
+
+function PostFooter(props: PostFooterProps) {
+  const [likes] = useState<number>(props.likes);
```
Comment (marvindv) [inline frontend/src/component/PostFooter.tsx]: Use props directly.

File: frontend/src/component/PostHeader.tsx [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+import './Components.css';
+
+interface PostHeaderProps {
+  name: string;
+  time: string;
+  isComment: boolean;
+}
+
+function PostHeader(props: PostHeaderProps) {
+  const [name] = useState<string>(props.name);
```
Comment (marvindv) [inline frontend/src/component/PostHeader.tsx]: Use props directly.

File: frontend/src/component/PostHeader.tsx [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+  const [time] = useState<string>(props.time);
+
+  var className = props.isComment ? 'Comment-Header' : 'Post-Header';
```
Comment (marvindv) [inline frontend/src/component/PostHeader.tsx]: Use `const`.

File: frontend/src/index.tsx [hunk 12]
```diff
@@ -1,17 +1,32 @@
...
 import ReactDOM from 'react-dom';
-import App from './App';
+import Post from './component/Post';
 import './index.css';
-import reportWebVitals from './reportWebVitals';
 
 ReactDOM.render(
   <React.StrictMode>
-    <App />
+    <div className='Center'>
```
Comment (marvindv) [inline frontend/src/index.tsx]: It is common to have an `App` component. In `index.tsx` is usually all that stuff like the strict mode and required wrapper components for things like Redux and in these wrappers one adds the `App` component. Then in the `App` component there is the actual root component from a layout point of view.

File: frontend/src/index.tsx [hunk 13]
```diff
@@ -1,17 +1,32 @@
...
+      />
+    </div>
   </React.StrictMode>,
   document.getElementById('root')
 );
 
 // If you want to start measuring performance in your app, pass a function
 // to log results (for example: reportWebVitals(console.log))
 // or send to an analytics endpoint. Learn more: https://bit.ly/CRA-vitals
-reportWebVitals();
```
Comment (marvindv) [inline frontend/src/index.tsx]: Why did you remove this line?

File: frontend/src/component/Components.css [hunk 14]
```diff
@@ -0,0 +1,89 @@
+/* Post */
+
+.Post-Header {
+  color: white;
+  font-size: larger;
+  background-color: #204e4a;
+  border: 2pt solid #204e4a;
```
Comment (marvindv) [inline frontend/src/component/Components.css]: Use `px` unit instead of `pt` or do you have a particular reason to use `pt`?

**Review by marvindv — CHANGES_REQUESTED** [review]

File: frontend/.vscode/settings.json (hunk 1 above)
Comment (ls505-hro) [reply frontend/.vscode/settings.json]: No, VS-Code asked me for this formatter. In my first commits I changed it back before pushing but it seems to have gone through in my later commits.

File: frontend/src/App.tsx (hunk 2 above)
Comment (ls505-hro) [reply frontend/src/App.tsx]: I have reverted to prettier now but every time I save it will format the imports in the same order as here. I also tried setting up prettier as my default formatter but the outcome is the same. Have you maybe forgotten to format this file beforehand?

File: frontend/src/index.tsx (hunk 13 above)
Comment (ls505-hro) [reply frontend/src/index.tsx]: I cant remember removing it. Probably a mistake.

File: frontend/src/App.tsx (hunk 2 above)
Comment (marvindv) [reply frontend/src/App.tsx]: Yes of course, like mentioned in the comment below. So nothing to see here

**Review by marvindv — APPROVED** [review]
